draft-ietf-mls-protocol.md

```diff
@@ -1290,70 +1290,150 @@ Given these inputs, the derivation of secrets for an epoch
+| `exporter_secret`       | "exporter"    |
+| `confirmation_key`      | "confirm"     |
+| `authentication_secret` | "stateauth"   |
+| `recovery_secret`       | "recovery"    |
 
 ## Pre-Shared Keys
 
 Groups which already have an out-of-band mechanism to generate
-shared group secrets can inject those in the MLS key schedule to seed
+shared group secrets can inject those into the MLS key schedule to seed
 the MLS group secrets computations by this external entropy.
 
-At any epoch, including the initial state, an application can decide
-to synchronize the injection of a PSK into the MLS key schedule.
-
-This mechanism can be used to improve security in the cases where
-having a full run of updates across members is too expensive or in
-the case where the external group key establishment mechanism provides
+The injecting of an external PSK can improve security in the cases
+where having a full run of updates across members is too expensive or in the
+case where the external group key establishment mechanism provides
 stronger security against classical or quantum adversaries.
 
-The security level associated with the PSK injected in the key schedule
-SHOULD match at least the security level of the ciphersuite in use in
-the group.
-
 Note that, as a PSK may have a different lifetime than an update, it
 does not necessarily provide the same Forward Secrecy (FS) or Post-Compromise
 Security (PCS) guarantees as a Commit message.
 
-<!-- OPEN ISSUE: We have to decide if we want an external coordination
-via the application of a Handshake proposal. -->
+<!-- OPEN ISSUEClarify lifetime vs security level mandated above. E.g. if
+the PSK security expires before the next update (shorter PSK lifetime than
+update), does that constitute a weaker security level -->
+
+<!-- OPEN ISSUE Define "security level", and what it means to match the
+security level of the ciphersuite used in the group. -->
+
+Each PSK in MLS has a type that designates how it was provisioned.
+External PSKs are provided by the application logic. Other resumption PSKs
+are derived from the MLS key schedule, and used in cases where it is
+necessary to authenticate a member's participation in a prior group state.
+In particular, in addition to external PSK types, a PSK derived from within MLS
+may be used in the following cases:
+
+  - Re-Initialization: If, during the lifetime of the group, a change in the
+    fixed group parameters becomes necessary, e.g. if the ciphersuite used
+    by the group is deprecated or if the protocol version should be upgraded,
+    a PSK can be used to carry entropy from the old group forward into a new
+    group with the desired parameters.
+
+  - Recovery: If the group state of one or more members of the group deviates
+    from the rest, they can be re-added to the group such that participants who
+    were group members at the time of deviation and are still in the group can
+    verify that the added member was previously in the group.
```

Comment:
Please elaborate on how recovery happens. From skimming, it's not clear how I would do this

Reply:
Good point. That's a relic from a past design that actually supported recovery. I'll remove it, as it is not relevant anymore.